detection_rules/schemas/base.py

```diff
@@ -0,0 +1,129 @@
+
+    @classmethod
+    @cached
+    def get_schema(cls, role=jsl.DEFAULT_ROLE, ordered=False):
+        """Wrap jsl.Document.get_schema to add caching."""
+        return super(GenericSchema, cls).get_schema(role=role, ordered=ordered)
+
+    @classmethod
+    @cached
+    def validate(cls, document, role=None):
+        """Validate a document against this schema."""
+        schema = cls.get_schema(role=role)
+        return jsonschema.validate(document, schema)
+
+
+class TomlMetadata(GenericSchema):
+    """Schema for siem rule toml metadata."""
+
+    creation_date = jsl.StringField(required=True, pattern=DATE_PATTERN, default=time.strftime('%Y/%m/%d'))
+
+    # added to query with rule.optimize()
+    # rule validated against each ecs schema contained
+    ecs_version = jsl.ArrayField(
+        jsl.StringField(pattern=VERSION_PATTERN, required=True, default=ecs.get_max_version()), required=True)
+    maturity = jsl.StringField(enum=MATURITY_LEVELS, default='development', required=True)
+
+    # if present, add to query
+    os_type_list = jsl.ArrayField(jsl.StringField(enum=OS_OPTIONS), required=False)
+    related_endpoint_rules = jsl.ArrayField(jsl.ArrayField(jsl.StringField(), min_items=2, max_items=2),
+                                            required=False)
+    updated_date = jsl.StringField(required=True, pattern=DATE_PATTERN, default=time.strftime('%Y/%m/%d'))
+
+
+class BaseApiSchema(GenericSchema):
+    """Base API schema with generic methods."""
+
+    STACK_VERSION = str()
+
+    rule_id = jsl.StringField(pattern=UUID_PATTERN, required=True)
+    type = jsl.StringField(required=True)
+
+    @classmethod
+    @cached
+    def versioned(cls):
+        """Get a subclass that is version aware."""
+        attrs = {"version": jsl.IntField(minimum=1, default=1, required=True)}
+        return type("Versioned" + cls.__name__, (cls, ), attrs)
+
+    @classmethod
+    def validate(cls, document, role=None, toml=False):
+        """Validate a document against this API schema."""
+        if toml:
+            role = role or document.get("rule", {}).get("type")
+            return cls.toml_schema().validate(document, role=role)
+
+        role = role or document.get("type")
+        return super(BaseApiSchema, cls).validate(document, role=role)
+
+    @classmethod
+    @cached
+    def markdown_fields(cls, role=None):
+        properties = cls.get_schema(role)["properties"]
+        return {p for p in properties if properties[p].get("format") == "markdown"}
+
+    @classmethod
+    @cached
+    def toml_schema(cls):
+        """Create a custom TOML schema class that includes this API schema."""
+        attrs = {
+            "metadata": jsl.DocumentField(TomlMetadata, required=True),
+            "rule": jsl.DocumentField(cls, required=True)
+        }
+        return type("Versioned" + cls.__name__, (GenericSchema, ), attrs)
+
+    @classmethod
+    def downgrade(cls, target_cls, document, role=None):
```

Comment:
```suggestion
    def coerce_by_release(cls, release_cls, document, role=None):
```

Can you rename it to this or similar since it is more of a coercion conversion than a downgrade. As an example, downloading a 7.9 rule from Kibana will need to be _coerced_ to be compatible with this repo due to needing to strip API only fields, by using this function

Reply:
I'm not sure what you mean with your distinction between downgrade vs coerce. I think downgrade is actually the right word here. Our approach is actually very similar to [alembic](https://alembic.sqlalchemy.org/en/latest/tutorial.html#downgrading), which is used to track SQL schemas over time so you can upgrade and downgrade between schema changes, step by step.

Also I'm expecting this to be used more going the other way: downgrade a rule before uploading it to an old version of Kibana.

Reply:
I'm not tied to any particular name, it just seems inaccurate that _making a 7.9 rule downloaded from Kibana compatible with the repo 7.9 schema_ is called a `downgrade`.

🤷 not a blocker for me regardless, so leave it if you feel that strongly about it

Reply:
Thanks for the changes